### ISSUE TITLE
Prelaunch tasks without a command don't work the second time

### ISSUE THREAD

**Issue by s73obrien** [microsoft/vscode#92792]
Issue Type: <b>Bug</b>

Repro repo here: https://github.com/s73obrien/task-tracking-repro .

- Clone repo and run "Launch Program" twice (2x).
- The second time, after about 8-10 seconds, it will error out with 'The specified task cannot be tracked'.
- Please note that this happens only *after* the first time running after a fresh VSCode instance is started.
- I used ping here just for reproduction purposes.
- This started showing up directly after the February 2020 update.
- I can confirm that this reproduces when using only shell-based tasks (no npm involved) as well.

VS Code version: Code 1.43.0 (78a4c91400152c0f27ba4d363eb56d2835f9903a, 2020-03-09T19:47:57.235Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-4910MQ CPU @ 2.90GHz (8 x 2893)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.94GB (21.99GB free)|
|Process Argv|C:\Users\sobrien\Documents\task-tracking-repro|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (15)</summary>

Extension|Author (truncated)|Version
---|---|---
pascal|ale|9.0.0
pascal-formatter|ale|2.2.0
JSDocTagComplete|Hoo|0.0.2
innosetup|idl|1.2.0
debugger-for-chrome|msj|4.12.6
show-offset|ram|0.0.4
java|red|0.58.0
vscode-xml|red|0.10.1
addDocComments|ste|0.0.8
vscodeintellicode|Vis|1.2.6
vscode-java-debug|vsc|0.25.1
vscode-java-dependency|vsc|0.9.0
vscode-java-pack|vsc|0.8.1
vscode-java-test|vsc|0.22.2
vscode-maven|vsc|0.21.1

(1 theme extensions excluded)

</details>

**Comment by s73obrien** [comment]
I have also reproduced this without any extensions enabled

**Comment by Feynstein** [comment]
I also have the same issue which seems to have re-appeared with the last update. It seems to be linked to issue #63705 (and multiple others linked to it) which is closed atm. I tried adding "isBackground" to my sub-tasks but nothing seems to work. I currently bypass the problem by compiling outside of vscode. It might be the version tag that's at fault, but I recently switched to vscode and I'm not fully familiar.

tasks.json
```
 {
    "version": "2.0.0",
    "options": {
        "cwd": "${workspaceRoot}"
    },
    "tasks": [
        {
            "label": "cmake",
            "args": ["-DCMAKE_BUILD_TYPE=Debug"],
            "type": "shell",
            "command": "cmake",
            "isBackground": true
        },
        {
            "label": "make",
            "args": ["-j16"],
            "type": "shell",
            "command": "make",
            "isBackground": true
        },
        {
            "label": "Build",
            "dependsOn": ["cmake", "make"],
            "group":"build",
        }
    ]
}
```
launch.json
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "cwd": "${fileDirname}"
        },
        {
            "name": "(gdb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": ".../bin/program",
            "args": ["--loglevel=2", "--su"],
            "stopAtEntry": false,
            "cwd": ".../bin/",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "preLaunchTask": "Build",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": false
                },
                {
                    "description": "Enable only unhandled exceptions",
                    "text": "catch throw terminate",
                    "ignoreFailures": false
                }
            ],
        
            
        }
    ]
}
```

**Comment by alexr00** [comment]
@s73obrien thank you for the very simple example! I debugged it, and it looks like we are literally taking longer to run the tasks the second time, and because of that the debug system times out. For example, if you change your npm task to only ping 9 times instead of 12, it makes it before the debug system timeout. That's not a solution, it just illustrates what's going on.

Clearly the task system should try not to take noticeably longer to run tasks the second time; however, if you do have a long running task you could hit the debug system task timeout anyway. I'll look into what could be slowing things down so much the second time.

If you're going to have non-background tasks as your prelaunch task, they must take less than 10 seconds to run, otherwise you need to convert them into background tasks.

@isidorn FYI

**Comment by isidorn** [comment]
Idealy we could also ask the user from debug land if after 10 seconds he wants to wait more..

**Comment by Feynstein** [comment]
@alexr00 Can you refer me to an issue that's more related to my problem? So that I don't file a new duplicate for the background task system not working and timing out my debug session each time I change my code? Now that you have changed the name of the issue, my comment became irrelevant...

**Comment by s73obrien** [comment]
@alexr00 For clarification:
Does the use of ```dependsOn``` force one's tasks to be run non-background?  There seems to be precious little documentation about the expected behavior (unless I am missing it).
Or more to the point: why is it that, when run on its own, ```npm: task2``` can seemingly run as long as it pleases?

@Feynstein have you tested to see if your tasks will run successfully once on a fresh instantiation of VSCode and then not-so-successfully thereafter, as described above?  Your tasks.json looks identical to mine, topologically-speaking.

**Comment by Feynstein** [comment]
@s73obrien Oh yeah... I closed vscode, then clean build then reopenend and started debug and even though it was long, it did work! Then I changed a line, so there was basically only linking to do and it timed out! I'm in the right issue after all!

**Comment by alexr00** [comment]
@s73obrien the debug system defines that a non-background task needs to finish in 10 seconds. As @isidorn said, ideally we'd ask the user after 10 seconds. You can have a background task that uses `dependsOn` (but the task in the depends on can't be background https://github.com/microsoft/vscode/issues/70283), but to be a background task that can be used as a prelaunch task you must have a watching problem matcher.

**Comment by Feynstein** [comment]
@alexr00 But as I mentionned in my last comment, on a clean boot of vscode running my whole compilation (after a make clean) doesn't trigger any timeout. And then changing one line in one file (which is basically at least 10x faster) I get a timeout. This is unstable behaviour if I ever saw any.

**Comment by alexr00** [comment]
Yes: 

> Clearly the task system should try not to take noticeably longer to run tasks the second time; however, if you do have a long running task you could hit the debug system task timeout anyway. I'll look into what could be slowing things down so much the second time.

**Comment by bwateratmsft** [comment]
A 10 second time limit seems _severely_ too restrictive to me. Even on a small .NET Core project, `dotnet build` can take 10+ seconds if NuGet restore is slow. In almost any reasonable Node.js project I can't imagine `npm install` taking _less_ than 10 seconds.

According to that rule, `dotnet build` should become a background task, but because of #70283 that makes Docker debugging impossible, because `docker build` has to depend on `dotnet build`.

I think we should rethink that 10 second timeout. Maybe 10 seconds without any output from the task, but even that is seriously restrictive.

**Comment by alexr00** [comment]
The 10 second timeout ended up being a red herring. There is a 10 second timeout, but is on the task _starting_ not completing. Fix is pushed to master and will be available in the text insiders.

**Comment by bwateratmsft** [comment]
Ah, that makes sense. Ten seconds to start is reasonable.

**Comment by haniamr** [comment]
> The 10 second timeout ended up being a red herring. There is a 10 second timeout, but is on the task _starting_ not completing. Fix is pushed to master and will be available in the text insiders.

Thanks Alex, is it going to be available in the Insiders nightly build today? Also, are there any plans to have a servicing release for the stable branch?

**Comment by alexr00** [comment]
Yes, it is in the latest insiders now. It is also planned to be included in the 1.43.1 sable release.
